### PR TITLE
ported to node

### PR DESCRIPTION
**Pull request by CrashInfinite**
This ports the elm-static-site python script to node.

-- To Do --
- ~~fix executeBash function to run ./runner.sh from inside node file instead of external run file~~
- ~~fix path issues (remove full path)~~
- Would be good to check for folders instead of throwing errors (e.g. output folder already exists)
- ~~Refactor and tidy code in general~~

### CODE REVIEW

File: js/convert.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
+var fs = require('fs');
+var glob = require('glob');
+var linebyline = require('n-readlines');
+var cmd = require('node-cmd');
+var exec = require('child_process').exec;
```
Comment (eeue56) [inline js/convert.js]: Use https://www.npmjs.com/package/cross-spawn instead, because `child_process` breaks on Windows

File: js/convert.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+
+var path = '/Users/psuc44/dev/shaun/examples/';
```
Comment (eeue56) [inline js/convert.js]: Remove hardcoded path

File: js/convert.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    var filelist = [];
+    var files =  glob.sync(path + '**').filter(isFile);
+
+    for (var i = 0; i < files.length; i++) {
+        if (getFilename(files[i]).split('.')[1] == 'elm') {
+            if (hasView(files[i]) == true) { 
+                filelist.push(files[i]);
+            }
+        }
+    }
```
Comment (eeue56) [inline js/convert.js]: You can convert this block (14-20) into

``` javascript
var fileList = files.filter(function(file){
  return getFilename(file).split('.')[1] === 'elm';
}).filter(hasView);
```

File: js/convert.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    return `import ${module_name}`
+}
+
+function portName(module_name) {
+    return module_name.replace('.','').toLowerCase();
+}
+
+function fileName(module_name, basedir, with_lower) {
+
+    if (basedir == null) {
```
Comment (eeue56) [inline js/convert.js]: `=== null`

File: js/convert.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        basedir = './';
+    } 
+    var with_basedir = basedir + module_name.replace('.','/');
+
+    if (with_lower == true) {
```
Comment (eeue56) [inline js/convert.js]: `if (with_lower)`

File: js/convert.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    } return with_basedir;
+}
+
+function generateMapping(port, file) {
+    return `echo "fs.writeFile('${file}.html', elm.ports['${port}']);" >> _main.js`
+}
+
+function generate_vdom(module_names, basedir) {
+
+    var port_files = []
```
Comment (eeue56) [inline js/convert.js]: semicolon

File: js/convert.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            'port' : portName(module_names[i]),
+            'filename' : fileName(module_names[i], basedir, false)
+        });
+    }
+
+    var ports = module_names.map(generatePort).join('\n');
+    var imports = module_names.map(generateImport).join('\n');
+    var maps = [];
+
+    var port_file_values = []
```
Comment (eeue56) [inline js/convert.js]: semicolon

File: js/convert.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    
+    for (var i = port_files.length-1; i >= 0; i--) {
+
+        var curr = port_files[i];
+
+        maps.push(generateMapping(curr.port, curr.filename));
+
+        port_file_values.push(curr.filename)
+
+    }
```
Comment (eeue56) [inline js/convert.js]: this block could be `var port_file_values = port_files.map(function(curr){...})` instead

File: js/convert.js [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+    // makeExecutable(runner_filename);
+
+    executeBash(runner_filename);
+
+}
+
+function makeFolders(filenames) {
+
+    for (var i = 0; i < filenames.length; i++) {
```
Comment (eeue56) [inline js/convert.js]: `filenames.forEach(function(filename){...})`

File: js/convert.js [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+}
+}
+
+function isFile(path) {
+    return fs.lstatSync(path).isFile();
+}
+
+function getFilename(path) {
+    var parts = path.split('/');
+    return parts[parts.length-1];
```
Comment (eeue56) [inline js/convert.js]: space before and after `-`, e.g `parts.length - 1`

File: js/convert.js [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+function hasView(filename) {
+    var liner = new linebyline(filename);
+    var line;
+
+    while (line = liner.next()) {
+        if (line.indexOf('view =') != -1) {
+            return true;
+        }
+    } return false;
```
Comment (eeue56) [inline js/convert.js]: this should be on a line by itself

File: js/run [hunk 12]
```diff
@@ -0,0 +1,3 @@
+node cleanStatic.js
+./runner.sh
+
```
Comment (eeue56) [inline js/run]: should be saved as `run.sh` instead

**Comment by eeue56** [comment]
Some minor cleanup notes, but thanks a lot for this!

File: js/convert.js (hunk 11 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 10 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 7 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 6 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 5 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 4 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 2 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 9 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/run (hunk 12 above)
Comment (CrashInfinite) [inline js/run]: done

File: js/convert.js (hunk 8 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+      return getFilename(file).split('.')[1] === 'elm';
+    }).filter(hasView).map(function(file) {
+        return file.replace(path,'');
+    });
+
+    // files.forEach(function(filename) {
+    //     filename.replace(path, '');
+    // })
+
+    console.log('for2', files);
```
Comment (eeue56) [inline js/convert.js]: kill the console.log

File: js/convert.js [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+    }   
+    console.log('for', filelist);
+
+    var files = files.filter(function(file){
+      return getFilename(file).split('.')[1] === 'elm';
+    }).filter(hasView).map(function(file) {
+        return file.replace(path,'');
+    });
+
+    // files.forEach(function(filename) {
```
Comment (eeue56) [inline js/convert.js]: kill the dead code

File: js/convert.js [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+    var files =  glob.sync(path + '**').filter(isFile);
+
+    for (var i = 0; i < files.length; i++) {
+        if (getFilename(files[i]).split('.')[1] == 'elm') {
+            if (hasView(files[i]) == true) { 
+                filelist.push(files[i].replace(path,''));
+            }
+        }
+    }   
+    console.log('for', filelist);
```
Comment (eeue56) [inline js/convert.js]: console.log

File: js/convert.js [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+    
+    // for (var i = port_files.length-1; i >= 0; i--) {
+
+    //     var curr = port_files[i];
+
+    //     maps.push(generateMapping(curr.port, curr.filename));
+
+    //      // port_file_values.push(curr.filename)
+
+    // }
```
Comment (eeue56) [inline js/convert.js]: dead code

File: js/convert.js (hunk 13 above)
Comment (CrashInfinite) [inline js/convert.js]: dead

File: js/convert.js (hunk 14 above)
Comment (CrashInfinite) [inline js/convert.js]: You can't kill the dead.... but it's gone

File: js/convert.js (hunk 15 above)
Comment (CrashInfinite) [inline js/convert.js]: gone

File: js/convert.js (hunk 16 above)
Comment (CrashInfinite) [inline js/convert.js]: gone

File: js/convert.js [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+    var filelist = [];
+    var files =  glob.sync(path + '**').filter(isFile);
+
+    var files = files.filter(function(file){
+      return getFilename(file).split('.')[1] === 'elm';
+    }).filter(hasView).map(function(file) {
+        return file.replace(path,'');
+    });
+
+    return files
```
Comment (eeue56) [inline js/convert.js]: semicolon

File: js/convert.js [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+var exec = require('child_process').exec;
+
+function listFiles(path) {
+    var filelist = [];
+    var files =  glob.sync(path + '**').filter(isFile);
+
+    var files = files.filter(function(file){
+      return getFilename(file).split('.')[1] === 'elm';
+    }).filter(hasView).map(function(file) {
+        return file.replace(path,'');
```
Comment (eeue56) [inline js/convert.js]: space after comma

File: js/convert.js [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+function fileName(module_name, basedir, with_lower) {
+
+    if (basedir === null) {
+        basedir = './';
+    } 
+    var with_basedir = basedir + module_name.replace('.','/');
+
+    if (with_lower) {
+        return with_basedir.toLowerCase();
+    } return with_basedir;
```
Comment (eeue56) [inline js/convert.js]: put return statement on newline here

File: js/convert.js [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+            'filename' : fileName(module_names[i], basedir, false)
+        });
+    }
+
+    var ports = module_names.map(generatePort).join('\n');
+    var imports = module_names.map(generateImport).join('\n');
+
+    var port_file_values = port_files.map(function(curr) {
+        return curr.filename;
+    })
```
Comment (eeue56) [inline js/convert.js]: semicolon

File: js/convert.js [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+
+    var maps = port_files.map(function(curr) {
+        return generateMapping(curr.port, curr.filename);
+    })
```
Comment (eeue56) [inline js/convert.js]: semicolon

File: js/convert.js [hunk 22]
```diff
@@ -0,0 +1,169 @@
...
+    //for (var i = 0; i < filenames.length; i++) {
+    filenames.forEach(function(filename) {
+        if (filename.split('/').length > 1 || filename.startsWith('.') != -1) {
+            var dir = filename.substring(0, filename.lastIndexOf('/'));
+            try {
+                fs.mkdir(dir);
+            } catch (err) {
+                console.log(err);
+            }
+    }
```
Comment (eeue56) [inline js/convert.js]: this should be ident one more in order to match the opening if

File: js/convert.js [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+    var line;
+
+    while (line = liner.next()) {
+        if (line.indexOf('view =') != -1) {
+            return true;
+        }
+    } 
+
+    return false;
+
```
Comment (eeue56) [inline js/convert.js]: kill this empty line

File: js/convert.js (hunk 17 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 18 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 19 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 20 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 21 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 22 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 23 above)
Comment (CrashInfinite) [inline js/convert.js]: dead

File: js/convert.js [hunk 24]
```diff
@@ -0,0 +1,164 @@
+var fs = require('fs');
+var glob = require('glob');
+var linebyline = require('n-readlines');
+var cmd = require('node-cmd');
+var spawn = require('cross-spawn')
```
Comment (eeue56) [inline js/convert.js]: semicolon

File: js/convert.js [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+    return files;
+}
+
+function generatePort(module_name) {
+    var port_name = portName(module_name);
+    return `
+port ${port_name} : String 
+port ${port_name} =
+    ${module_name}.view 
+        |> render`
```
Comment (eeue56) [inline js/convert.js]: semicolon

File: js/convert.js [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+function generateImport(module_name) {
+    return `import ${module_name}`
```
Comment (eeue56) [inline js/convert.js]: semicolon

File: js/convert.js [hunk 27]
```diff
@@ -0,0 +1,164 @@
...
+    var with_basedir = basedir + module_name.replace('.','/');
+
+    if (with_lower) {
+        return with_basedir.toLowerCase();
+    } 
+    return with_basedir;
+}
+
+function generateMapping(port, file) {
+    return `echo "fs.writeFile('${file}.html', elm.ports['${port}']);" >> _main.js`
```
Comment (eeue56) [inline js/convert.js]: semi-colon

File: js/convert.js [hunk 28]
```diff
@@ -0,0 +1,164 @@
...
+function generate_vdom(module_names, basedir) {
+
+    var port_files = [];
+
+    for (var i = module_names.length - 1; i >= 0; i--) {
+        port_files.push({
+            'port' : portName(module_names[i]),
+            'filename' : fileName(module_names[i], basedir, false)
+        });
+    }
```
Comment (eeue56) [inline js/convert.js:63]: this could be a `map` but nbd

File: js/convert.js [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+
+function makeFolders(filenames) {
+
+    filenames.forEach(function(filename) {
+        if (filename.split('/').length > 1 || filename.startsWith('.') != -1) {
+            var dir = filename.substring(0, filename.lastIndexOf('/'));
+            try {
+                fs.mkdir(dir);
+            } catch (err) {
+                console.log(err);
```
Comment (eeue56) [inline js/convert.js]: `console.error` instead

File: js/convert.js [hunk 30]
```diff
@@ -0,0 +1,164 @@
...
+
+}
+
+function main() {
+    var files = listFiles('examples/').map(cleanUp);
+
+    generate_vdom(files, 'output/');
+}
+
+main()
```
Comment (eeue56) [inline js/convert.js]: semicolon

File: js/convert.js (hunk 27 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 26 above)
Comment (CrashInfinite) [inline js/convert.js]: done

**Comment by eeue56** [comment]
Awesome! Last few comments and then I'll merge

File: js/convert.js (hunk 25 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 24 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 29 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 30 above)
Comment (CrashInfinite) [inline js/convert.js]: done

File: js/convert.js (hunk 28 above)
Comment (CrashInfinite) [inline js/convert.js:63]: nbd :+1:

**Pull request merged**